src/org/traccar/protocol/Arnavi4FrameDecoder.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2017 Ivan Muratov (binakot@gmail.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.traccar.protocol;
+
+import org.jboss.netty.buffer.ChannelBuffer;
+import org.jboss.netty.channel.Channel;
+import org.jboss.netty.channel.ChannelHandlerContext;
+import org.jboss.netty.handler.codec.frame.FrameDecoder;
+
+public class Arnavi4FrameDecoder extends FrameDecoder {
```

Comment:
I have read the code once again and I believe it's completely wrong.

Biggest problem here is that you should never rely on buffer size. The whole purpose of frame decoder is to split stream of data into frames.

1. Am I understanding correctly that first message is always header? If the size is fixed, you should just read first 10 bytes as a header and save a flag somewhere to indicate that header has been read.

2. I believe regular packages should be split using length field. Never return original buffer from frame decoder. You need to extract frame using `readBytes` and send it upstream.